src/main/java/com/commercetools/sync/customers/CustomerSync.java

```diff
@@ -0,0 +1,235 @@
+     * @return a {@link CompletionStage} which contains an empty result after execution of the update
+     */
+    @Nonnull
+    private CompletionStage<Void> syncBatch(
+        @Nonnull final Set<Customer> oldCustomers,
+        @Nonnull final Set<CustomerDraft>  newCustomerDrafts) {
+
+        final Map<String, Customer> oldCustomerMap = oldCustomers
+            .stream()
+            .collect(toMap(Customer::getKey, identity()));
+
+        return CompletableFuture.allOf(newCustomerDrafts
+            .stream()
+            .map(customerDraft ->
+                referenceResolver
+                    .resolveReferences(customerDraft)
+                    .thenCompose(resolvedCustomerDraft -> syncDraft(oldCustomerMap, resolvedCustomerDraft))
+                    .exceptionally(completionException -> {
+                        final String errorMessage = format(FAILED_TO_PROCESS, customerDraft.getKey(),
+                            completionException.getMessage());
+                        handleError(new SyncException(errorMessage, completionException), 1);
+                        return null;
+                    })
+            )
+            .map(CompletionStage::toCompletableFuture)
+            .toArray(CompletableFuture[]::new));
+    }
+
+    @Nonnull
+    private CompletionStage<Void> syncDraft(
+        @Nonnull final Map<String, Customer> oldCustomerMap,
+        @Nonnull final CustomerDraft newCustomerDraft) {
+
+        final Customer oldCustomer = oldCustomerMap.get(newCustomerDraft.getKey());
+        return Optional.ofNullable(oldCustomer)
+                       .map(customer -> buildActionsAndUpdate(oldCustomer, newCustomerDraft))
+                       .orElseGet(() -> applyCallbackAndCreate(newCustomerDraft));
+    }
+
+    @SuppressFBWarnings("NP_NONNULL_PARAM_VIOLATION") // https://github.com/findbugsproject/findbugs/issues/79
+    @Nonnull
+    private CompletionStage<Void> buildActionsAndUpdate(
+        @Nonnull final Customer oldCustomer,
+        @Nonnull final CustomerDraft newCustomerDraft) {
+
+        // todo (ahmetoz) no action utils, yet.
+        return completedFuture(null);
+    }
+
+    @Nonnull
+    private CompletionStage<Void> applyCallbackAndCreate(@Nonnull final CustomerDraft customerDraft) {
+
+        return syncOptions
+            .applyBeforeCreateCallback(customerDraft)
+            .map(draft -> customerService
+                .createCustomer(draft)
+                .thenAccept(customerOptional -> {
+                    if (customerOptional.isPresent()) {
+                        statistics.incrementCreated();
+                    } else {
+                        statistics.incrementFailed();
+                    }
+                }))
+            .orElseGet(() -> CompletableFuture.completedFuture(null));
+    }
+
+    /**
+     * Given a {@link String} {@code errorMessage} and a {@link Throwable} {@code exception}, this method calls the
```

Comment:
I'm not sure if this comment is valid. Can you check?

Reply:
I removed some javadocs in here, for all private methods: https://github.com/commercetools/commercetools-sync-java/pull/590/commits/a94cacc484b1e7578de4d86291f6b44723e39876